### DISCUSSION TITLE
GamePad para PC

### DISCUSSION THREAD

**Discussion by EdmarMagalhaes** [NicoHood/HID#296] [Q&A]
Hi all right, I would like more information about gamepad because I'm researching how to make a flight controller with rotary encoder axes, I see that on the gamepad page it says I can create 4 16-bit axes (X, Y, RX, RY) , 28-bit axes (Z, RZ) , but I would like more information as I am a beginner in Arduino. I already installed HoodLoader2 with HID Project + gamepad my arduino uno already shows how game controller, I just need more info about gamepad. If you can help me with a more detailed example I would be very grateful! See you later.

**Comment by NicoHood** [comment]
Did you try the gamepad example?
https://github.com/NicoHood/HID/blob/master/examples/Gamepad/Gamepad.ino

**Comment by EdmarMagalhaes** [comment]
I've tried it with the gamepad example, it just doesn't work, it just
doesn't work with buttons or my rotary encoder, I don't think it's for
arduino uno 16u2. I would like to know how to get my arduino back as it was
before the changes, how it came from the factory, can you help me?

**Reply by NicoHood** [reply]
It should work with the 16u2 unless you do not use all usb endpoints. Meaning only a gamepad is supported, no other keyboard etc at the same time. you can read up on that here:
https://github.com/NicoHood/HID/wiki/API-Documentation

Here is how to get back to factor default:
https://github.com/NicoHood/HoodLoader2/wiki/Troubleshoot-FAQ#restore-stock-16u2-bootloader

**Reply by EdmarMagalhaes** [reply]
Hello Nico, how are you?
In the example of the gamepad where I physically connect a button and the axis to the arduino, my question is which pins I use?

**Reply by NicoHood** [reply]
I dont understand that question. The example uses pin 2:
https://github.com/NicoHood/HID/blob/master/examples/Gamepad/Gamepad.ino#L18

But you'd need to code your own logic of course

**Reply by EdmarMagalhaes** [reply]
This logic is for Joystick.h + Encoder.h! Can I use this logic with Gamepad.h + Encoder.h?

int pinos [] = {
  7,8,9,10,11,12};

  int lastButtonState[6] = {0,0,0,0,0,0};

  int pulsosVolta= 3111;
  
#include "Joystick.h"
#include <Encoder.h>

int ganho = 0;
int Position = 0;

long oldPosition  = -999;

Encoder myEnc(3, 2);

Joystick_ Joystick(JOYSTICK_DEFAULT_REPORT_ID, 
  JOYSTICK_TYPE_MULTI_AXIS, 15, 0,
  false, false, true, true,true, true,
  false, false, false, false, true);

void setup() {

//Serial.begin(115200);

  Joystick.setZAxisRange   (0, 1023);
  Joystick.setRxAxisRange  (0, 1023);
  Joystick.setRyAxisRange  (0, 1023);
  Joystick.setRzAxisRange  (0, 1023);
  Joystick.setSteeringRange(0, (pulsosVolta*2.8));
  
for(int i=0; i<6; i++){
  pinMode(pinos[i],INPUT_PULLUP);
  }

  pinMode(4,INPUT_PULLUP);
  pinMode(5,INPUT_PULLUP);
  pinMode(6,INPUT_PULLUP);
  pinMode(A5,INPUT_PULLUP);

    Joystick.begin();

pinMode(13,OUTPUT);
  analogWrite(13,255);
  
}

void loop() {

long newPosition = myEnc.read();

if (digitalRead(4)== LOW){
 ganho = (map(newPosition,(pulsosVolta*(-1.4)),(pulsosVolta*1.4),(pulsosVolta*1.4),(pulsosVolta*(-1.4))));
}


if (newPosition != oldPosition){
   oldPosition = newPosition;
}


Position = (newPosition + ganho);

if (Position >= pulsosVolta*1.4){
  Position = pulsosVolta*1.4;
}
if (Position <= pulsosVolta*(-1.4)){
  Position = pulsosVolta*(-1.4);
}

//Serial.print("embreagem    :");
//Serial.println(analogRead(A4));
//Serial.print("FREIO        :");
//Serial.println(analogRead(A1));
//Serial.print("ACELERADOR   :");
//Serial.println(analogRead(A2));


//Serial.print("GANHO:             ");
//Serial.println(ganho);
//Serial.print("POSIÇÃO SEM GANHO: ");
//Serial.println(newPosition);
//Serial.print("POSITION:          ");
//Serial.println(Position);

Joystick.setSteering (map(Position,pulsosVolta*(-1.4),pulsosVolta*1.4,0,pulsosVolta*2.8));
Joystick.setZAxis  (analogRead(A2));
Joystick.setRxAxis (analogRead(A1));
Joystick.setRyAxis (analogRead(A0));
Joystick.setRzAxis (analogRead(A3));

for (int index = 0; index < 6; index++)
  {
    int currentButtonState = map(digitalRead(index + 7),HIGH,LOW,LOW,HIGH);
    
    
    if (currentButtonState != lastButtonState[index])
    {
     Joystick.setButton(index+8, currentButtonState);
      lastButtonState[index] = currentButtonState;
      
    }
    
    Joystick.setButton(14, map(digitalRead(A5),HIGH,LOW,LOW,HIGH));
   if((analogRead(A4)<=555)&&(digitalRead(6)==LOW)){
  Joystick.setButton(0,HIGH);
}

   if((analogRead(A4)<=555)&&(digitalRead(5)==LOW)){
  Joystick.setButton(1,HIGH);
}
   if((analogRead(A4)>=556)&&(analogRead(A4)<=595)&&(digitalRead(6)==LOW)){
  Joystick.setButton(2,HIGH);
}
   if((analogRead(A4)>=556)&&(analogRead(A4)<=595)&&(digitalRead(5)==LOW)){
  Joystick.setButton(3,HIGH);
}
   if((analogRead(A4)>=596)&&(analogRead(A4)<=637)&&(digitalRead(6)==LOW)){
  Joystick.setButton(4,HIGH);
}
   if((analogRead(A4)>=596)&&(analogRead(A4)<=637)&&(digitalRead(5)==LOW)){
  Joystick.setButton(5,HIGH);
}
   if((analogRead(A4)>=638)&&(digitalRead(6)==LOW)){
  Joystick.setButton(6,HIGH);
}
   if((analogRead(A4)>=638)&&(digitalRead(5)==LOW)){
  Joystick.setButton(7,HIGH);
}

if((digitalRead(6)==HIGH)&&(digitalRead(5)==HIGH)){
  Joystick.setButton(0,LOW);
  Joystick.setButton(1,LOW);
  Joystick.setButton(2,LOW);
  Joystick.setButton(3,LOW);
  Joystick.setButton(4,LOW);
  Joystick.setButton(5,LOW);
  Joystick.setButton(6,LOW);
  Joystick.setButton(7,LOW);
}   
  }  
}

**Reply by NicoHood** [reply]
I am sorry, but I am not able to help with such detailed questions. Also `Joystick` is not part of this library, so I dont know what you are asking for here.
